---
title: "Differential Expression Analysis using edgeR"
author: "Chandan Pavuluri"
output: rmarkdown::github_document
---

```{r, warning= FALSE}
#Create a folder for the input files and results of the analysis
#Enter path of the folder that was created by you.
#at the end of path add "/".

path<-"C:/Users/ckpav/Documents/RNA-Seq_DEA_Pathway/edgeR/"

#Copy the input file in to the path
#Enter input file name of the raw counts file
Input<-"GSE150392.csv"
```


# Loading required packages
```{r, message=FALSE, warning=FALSE, results='hide'}
if(!require('BiocManager',quietly = T)) install.packages("BiocManager");library(BiocManager)
if(!require('edgeR',quietly = T))
  BiocManager::install("edgeR");library(edgeR)
if(!require('tidyverse',quietly = T)) install.packages("tidyverse");library(tidyverse)
if(!require('dplyr',quietly = T)) install.packages("dplyr");library(dplyr)
if(!require('EnhancedVolcano',quietly = T))
  BiocManager::install("EnhancedVolcano");library(EnhancedVolcano)

```


#Reading the raw counts(Counts_matrix) and creating metadata file
```{r, warning=FALSE}
# Reading the raw counts
counts_matrix <- read.table(paste0(path,Input),header=T,sep = ",")
dimnames(counts_matrix)[[1]] <- counts_matrix[,1]
counts_matrix <- counts_matrix[,-1]
dim(counts_matrix)

# Creating a metadata file 
meta_data<-data.frame(colnames(counts_matrix))
colnames(meta_data)[1] <- "Samples"

meta_data <- meta_data %>%
  mutate(Condition = if_else(str_detect(meta_data$Samples,"\\Cov"),"COVID","Control"))

Condition <- factor(meta_data[,2])

```


#Processing the counts data
```{r, warning=FALSE}
#Creating object for the counts table
dge <- DGEList(counts=counts_matrix,group=meta_data[,2])

#filtering by expression
keep <- filterByExpr(dge)
dge <- dge[keep,]

dim(dge)


#Normalization using Trimmed mean mvalues.
dge <- calcNormFactors(dge, method="TMM")

```


# MDS plot Generation 
```{r, warning=FALSE}

pch <- c(1,2)
colors <- rep(c("red", "blue"),1)
#png(paste0(path,"/MDS_",Input,"_edgeR.png"), 700, 500, pointsize=20)
plotMDS(dge, col=colors[factor(meta_data[,2])], pch=pch[factor(meta_data[,2])])
legend("topleft", legend=levels(factor(meta_data[,2])),pch=pch, col=colors, ncol=2)
title("MDS plot")
#dev.off()


```


# Modelling the data 
```{r, warning=FALSE}
design <- model.matrix(~0+factor(meta_data[,2]))
colnames(design)[1]<-gsub(".*)", "", colnames(design)[1])
colnames(design)[2]<-gsub(".*)", "", colnames(design)[2])


#estimates the Dispersion from maximum likelihood
dge<-estimateDisp(dge, design=design, robust = T)

#Fit a quasi-likelihood negative binomial generalized log-linear model to count data 
fit <- glmQLFit(dge, design)

#Choosing the contrast
compare<-"COVID-Control"

#Constructing custom matrix
contr<-makeContrasts(compare, levels=design)

#Count genewise statistical tests
qlf<-glmQLFTest(fit, contrast=contr)

#Taking top results of the statistical test and calculating FDR
edgeR_result<-topTags(qlf, n = 100000, adjust.method = "BH")
results<-edgeR_result$table
head(qlf$table)
```


# Volcano plot generator
```{r,warning=FALSE}

#png(paste0(path,"/Volcano_",Input,"_edgeR.png"), 700, 500, pointsize=20)
EnhancedVolcano(results,lab = NA,x = 'logFC',y = 'PValue',pCutoff = (10e-2)/2,FCcutoff = 1.0,  xlim = c(-5, 5),ylim = c(0, -log10(10e-12)))

 
```

# MD plot Generator
```{r,warning=FALSE}
#png(paste0(path,"/MD_",Input,"_edgeR.png"), 700, 500, pointsize=20)
plotMD(qlf,main=compare,p.value=0.01)
#dev.off()

```

# Filtering the results and writing to a CSV file and generating the stats of the pipeline.
```{r, warning=FALSE}
# Passing the results in to Dataframe 
A<-data.frame(results)

# Passing the rownames to a new column
A$Ensembl_ID<-row.names(A)

# Removing the remaining charcters after "."
A$Ensembl_ID <- gsub("\\..*","", row.names(A))

# Removing the duplicates
A<-unique(A)

# Changing the rownames to clean EnsemblIDs
row.names(A) <- A$Ensembl_ID

# Statistical filtering of the results
B<-A%>%
    filter(abs(logFC)>1)%>%
  arrange(desc(logFC))%>%
  filter(PValue<0.05)%>%
  filter(FDR<0.01)

# Creating the regulation column
B$Regulation <- ifelse(B$logFC>0, "Up", "Down")

# Passing the required columns to Dataframe
B1<-B%>%
  select(PValue,logFC,FDR,Regulation)
head(B1)

# writing the results to a csv
write.csv(B1,paste0(path,"edgeR_DE_",Input,".csv"))
```

# Comparing DESeq2 and edgeR results 
```{r, warning=FALSE, message=FALSE}
#DESeq2 file
Deseq2_results<-"DiffExp_DESeq2_GSE150392.csvsamples.csv"

#DEseq2 File pathway
Deseq2path<-"C:/Users/ckpav/Documents/RNA-Seq_DEA_Pathway/DESeq2_Pathway/"

#Reading the file
Deseq2<-read.table(paste0(Deseq2path,Deseq2_results),header=T,sep = ",")

#Extracting Deseq2 significantly differntially expressed genes
Deseq2<-data.frame(Deseq2[,1])
colnames(Deseq2)<-"EnsemblID"
dim(Deseq2)

#Extracting Deseq2 significantly differntially expressed genes
edgeR<-data.frame(row.names(B1))
colnames(edgeR)<-"EnsemblID"
dim(edgeR)

# Finding the common significantly differntially expressed genes
common_ID<-inner_join(Deseq2,edgeR)
dim(common_ID)
head(common_ID)
```

